ambry-store/src/main/java/com.github.ambry.store/DiskSpaceAllocator.java

```diff
@@ -0,0 +1,431 @@
+   * @throws IOException
+   */
+  private void addRequiredSegments(Map<Long, Long> overallRequirements) throws IOException {
+    for (Map.Entry<Long, Long> sizeAndSegmentsNeeded : overallRequirements.entrySet()) {
+      long sizeInBytes = sizeAndSegmentsNeeded.getKey();
+      long segmentsNeeded = sizeAndSegmentsNeeded.getValue();
+      while (reserveFiles.getCount(sizeInBytes) < segmentsNeeded) {
+        reserveFiles.add(sizeInBytes, createReserveFile(sizeInBytes));
+      }
+    }
+  }
+
+  /**
+   * Create and preallocate (if supported) a reserve file of the specified size.
+   * @param sizeInBytes the size to preallocate for the reserve file.
+   * @return the created file.
+   * @throws IOException if the file could not be created, or if an error occured during the fallocate call.
+   */
+  private File createReserveFile(long sizeInBytes) throws IOException {
+    File fileSizeDir = prepareDirectory(new File(reserveDir, FILE_SIZE_DIR_PREFIX + sizeInBytes));
+    File reserveFile;
+    do {
+      reserveFile = new File(fileSizeDir, generateFilename());
+    } while (!reserveFile.createNewFile());
+    Utils.preAllocateFileIfNeeded(reserveFile, sizeInBytes);
+    return reserveFile;
+  }
+
+  /**
+   * Create a directory if it does not yet exist and check that it is accessible.
+   * @param dir The directory to possibly create.
+   * @return the passed-in directory
+   * @throws IOException if the specified directory could not be created/read, or if a non-directory file with the same
+   *                     name already exists.
+   */
+  private static File prepareDirectory(File dir) throws IOException {
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+    if (!dir.isDirectory()) {
+      throw new IOException(dir.getAbsolutePath() + " is not a directory or could not be created");
+    }
+    return dir;
+  }
+
+  /**
+   * @param sizeInBytes the size of the files in this directory
+   * @return a directory name for this size. This is reserve_size_{n} where {n} is {@code sizeInBytes}
+   */
+  static String generateFileSizeDirName(long sizeInBytes) {
+    return FILE_SIZE_DIR_PREFIX + sizeInBytes;
+  }
+
+  /**
+   * @return a filename for a reserve file. This is reserve_ followed by a random UUID.
+   */
+  private static String generateFilename() {
+    return RESERVE_FILE_PREFIX + UUID.randomUUID();
+  }
+
+  /**
+   * Parse the file size from a file size directory name.
+   * @param fileSizeDirName the name of the file size directory.
+   * @return the parsed size in bytes, or {@code null} if the directory name was malformed.
+   */
+  private static Long getFileSizeForDirName(String fileSizeDirName) {
+    Long sizeInBytes = null;
+    if (fileSizeDirName.startsWith(FILE_SIZE_DIR_PREFIX)) {
+      String sizeString = fileSizeDirName.substring(FILE_SIZE_DIR_PREFIX.length());
+      try {
+        sizeInBytes = Long.parseLong(sizeString);
+      } catch (NumberFormatException ignored) {
```

Comment:
why do we ignore this?

Reply:
changed to not catch, since we should error out anyways if we find a directory that starts with the right prefix but has a malformed number after that